File: Problems/No.2615_Sum_of_Distances/sum_of_distances.cpp
class Solution {
public:
    vector<long long> distance(vector<int>& nums) {
        
        // parameters
        std::unordered_map<int, std::pair<int, long long>> num_count_sum;
        std::vector<long long> SoD(nums.size(), 0);

        // start computing
        for (int i=0; i<nums.size(); i++) {

            num_count_sum[nums[i]].first++;
            num_count_sum[nums[i]].second += i;

        }
        
        // calculate the distances
        std::unordered_map<int, std::pair<int, long long>> partial_NCS;
        for (long long i=0; i<nums.size(); i++) {

            int num = nums[i];
            auto [total_count, total_sum] = num_count_sum[num];
            
            partial_NCS[num].first++;
            partial_NCS[num].second += i;
            auto [left_count, left_sum] = partial_NCS[num];

            // ans for index i will be i * leftSize - prefixSum of left side + prefixSum of right side - i * rightSize
            SoD[i] = i*(2*left_count - total_count) - (2*left_sum - total_sum);

        } 

        return SoD;

    }
};

File: Problems/No.2615_Sum_of_Distances/sum_of_distances.py
class Solution(object):
    def distance(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """

        # parameters
        num_count_sum = {}
        SoD = [0] * len(nums)

        # start computing
        # count the total sum and # of each num
        for i in range(len(nums)):
            if nums[i] not in num_count_sum:
                num_count_sum[nums[i]] = [1, i]
            else:
                num_count_sum[nums[i]][0] += 1
                num_count_sum[nums[i]][1] += i
        
        # calculate the distances
        partial_NCS = {}
        for i in range(len(nums)):
            num = nums[i]
            [total_count, total_sum] = num_count_sum[num]
            
            if num not in partial_NCS:
                partial_NCS[num] = [1, i]
            else:
                partial_NCS[num][0] += 1
                partial_NCS[num][1] += i
            [left_count, left_sum] = partial_NCS[num]
           
            # ans for index i will be i * leftSize - prefixSum of left side + prefixSum of right side - i * rightSize
            SoD[i] = i*(2*left_count - total_count) - (2*left_sum - total_sum)

        return SoD
